File: ndb/03_dfs_bfs/3_2_bfs/test1/main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {
    static int n = 0;
    static int m = 0;
    static int[][] graph;
    // 상, 하, 좌, 우 정의
    static int[] dc = {-1, 1, 0, 0};
    static int[] dr = {0, 0, -1, 1};
    public static void main(String[] args) {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        try {
            // n, m 공백 구분 입력 받기
            String[] input = br.readLine().split(" ");
            n = Integer.parseInt(input[0]);
            m = Integer.parseInt(input[1]);

            // 2차원 리스트 맵 정보 입력 받기
            graph = new int[n][m];
            for (int i = 0; i < n; i++) {
                BufferedReader br2 = new BufferedReader(new InputStreamReader(System.in));
                String[] input2 = br2.readLine().split("");
                int[] row = new int[input2.length];
                for (int j = 0; j < input2.length; j++) {
                    row[j] = Integer.parseInt(input2[j]);
                }
                graph[i] = row;
            }

            System.out.println(bfs(0, 0)); // 정답 출력
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    // BFS 소스코드
    private static int bfs(int c, int r) {
        // Queue 구현을 위한 deque 라이브러리 사용
        Deque<int[]> queue = new ArrayDeque<>();
        int[] q = {c, r};
        queue.addFirst(q);

        // 큐가 빌 때까지 반복
        while (!queue.isEmpty()) {
            int[] que = queue.removeLast();
            c = que[0];
            r = que[1];

            // 현재 위치에서 네 방향으로 위치 확인
            for (int i = 0; i < 4; i ++) {
                int nc = c + dc[i];
                int nr = r + dr[i];

                // 미로 공간을 벗어난 경우 무시
                if (nc < 0 || nr < 0 || nc >= n || nr >=m) {
                    continue;
                }

                // 벽인 경우 무시
                if (graph[nc][nr] == 0) {
                    continue;
                }

                // 노드를 처음 방문하는 경우에만 최단거리 기록
                if (graph[nc][nr] == 1) {
                    graph[nc][nr] = graph[c][r] + 1;
                    int[] q2 = { nc, nr };
                    queue.addFirst(q2);

//                    for (int[] ints : queue) {
//                        System.out.println(Arrays.toString(ints));
//                    }
//                    System.out.println(Arrays.toString(graph[0]));
//                    System.out.println(Arrays.toString(graph[1]));
//                    System.out.println(Arrays.toString(graph[2]));
//                    System.out.println("---");
                }
            }
        }
        return graph[n - 1][m - 1];
    }
}

File: ndb/03_dfs_bfs/3_2_bfs/test1/main.dart
import 'dart:collection';
import 'dart:io';

void main(List<String> arguments) {
  // v(노드 개수), e(간선) 공백 구분하여 입력 받기
  var input1 = stdin.readLineSync()?.split(" ");
  if (input1 != null) {
    int v = int.parse(input1[0]);
    int e = int.parse(input1[1]);
    // 모든 노드에 대한 진입차수 0으로 초기화
    List<int> indegree = List.filled(v + 1, 0);
    // 각 노드에 연결된 간선 정보를 담기 위한 연결리스트(그래프) 초기화
    List<List<int>> graph = [];
    for (int i = 0; i < v + 1; i++) {
      graph.add([]);
    }
    // 방향 그래프의 간선 정보 입력받기
    for (int i = 0; i < e; i++) {
      var input = stdin.readLineSync()?.split(" ");
      if (input != null) {
        int a = int.parse(input[0]);
        int b = int.parse(input[1]);
        // 정점 A에서 B로 이동 가능
        graph[a].add(b);
        // 진입차수 1 증가
        indegree[b] += 1;
      }
    }
    print("graph");
    print(graph);
    print("indegree");
    print(indegree);
    // 위상 정렬
    List<int> result = [];
    // Queue 구현을 위한 라이브러리 사용
    final queue = Queue<int>();
    // 처음 시작할 때 진입차수가 0인 노드 큐에삽입
    for (int i = 1; i < v + 1; i++) {
      if (indegree[i] == 0) {
        queue.addFirst(i);
      }
    }
    // 큐가 빌때까지 반복
    while (queue.isNotEmpty) {
      // 큐에서 원소 꺼내기
      int now = queue.removeLast();
      result.add(now);
      print("pop");
      print(now);
      for (int i in graph[now]) {
        // 해당 원소와 연결된 노드의 진입차수 1 빼기
        indegree[i] -= 1;
        // 진입차수가 0이 되는 노드 큐에 삽입
        if (indegree[i] == 0) {
          queue.addFirst(i);
        }
      }
    }
    // 결과 출력
    print("result");
    print(result);
  }
}

/*
입력
5 6
1 2
2 3
3 4
2 5
1 5
5 4

출력
graph
[[], [2, 5], [3, 5], [4], [], [4]]
indegree
[0, 0, 1, 1, 2, 2]
pop
1
pop
2
pop
3
pop
5
pop
4
result
[1, 2, 3, 5, 4]
*/